Interleaved Randomized Benchmarking

ShellyGarion opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

#129

### Details and comments
See method https://arxiv.org/abs/1203.4550

ShellyGarion commented:
@dcmckayibm - could you please look at the updated `randomized_benchmarking_seq` function in: `circuit.py` and approve the API for interleaved RB?

ShellyGarion commented:
@dcmckayibm - here are examples of possible inputs `rb_opts` from the test:
```
Testing 1 qubit RB
{'nseeds': 3, 'length_vector': [1, 3, 4, 7], 'rb_pattern': [[0]], 'length_multiplier': 1, 'interleaved_gates': [['h 0', 'w 0', 'y 0']]}
{'nseeds': 3, 'length_vector': [1, 3, 4, 7], 'rb_pattern': [[0]], 'length_multiplier': [1], 'interleaved_gates': [['w 0', 'y 0']]}
Testing 2 qubit RB
{'nseeds': 3, 'length_vector': [1, 3, 4, 7], 'rb_pattern': [[0, 1]], 'length_multiplier': 1, 'interleaved_gates': [['h 1', 'v 1', 'cx 0 1', 'cx 1 0', 'v 0', 'y 0', 'z 1']]}
{'nseeds': 3, 'length_vector': [1, 3, 4, 7], 'rb_pattern': [[0, 1]], 'length_multiplier': [1], 'interleaved_gates': [['h 1', 'cx 0 1', 'w 1', 'x 1']]}
{'nseeds': 3, 'length_vector': [1, 3, 4, 7], 'rb_pattern': [[0], [1]], 'length_multiplier': 1, 'interleaved_gates': [['w 0'], ['h 0', 'w 0', 'y 0']]}
{'nseeds': 3, 'length_vector': [1, 3, 4, 7], 'rb_pattern': [[0], [1]], 'length_multiplier': [1, 2], 'interleaved_gates': [['x 0'], ['h 0', 'y 0']]}
```
In particular, do the last 2 examples (for simultaneous interleaved RB) make sense?
(note that both Cliffords are defined on 1 qubit, which is denoted by 0, but each one of them actually operates on a different qubit 0&1, as defined by the pattern)

dcmckayibm commented:
I think the interleaved gate should be a Clifford object.

I agree with the API. The interleave gate should be an n-qubit Clifford referenced to qubits 0,...,n-1. You need to add some error checking to make sure the right sized Clifford gets passed in.

ShellyGarion commented:
@dcmckayibm - could you please approve the API and code in `circuits.py` and `interleaved_fitters.py` ?
I also completed the tests of interleaved rb circuits in `rb_test.py`.
I still need to work on a Jupyter tutorial notebook, `test_fitters.py` and documentation.

ShellyGarion commented:
Uploaded a first version of a Jupyter notebook tutorial for interleaved RB

ShellyGarion commented:
@dcmckayibm - could you please look at the Jupyter notebook? 
How would you suggest to present the final output?

dcmckayibm commented on test/rb/test_rb.py:
```diff
@@ -83,7 +83,56 @@ def choose_multiplier(mult_opt, len_pattern):
+        '''
+        pattern_sizes = [len(pat) for pat in rb_pattern]
+        interleaved_gates = []
+        for (_, nq) in enumerate(pattern_sizes):
```
As we discussed let's not have a random test...pick a fixed interleaved sequence

ShellyGarion replied:
OK - done

ShellyGarion replied:
The fixed interleaved gates contain x gate on each qubit and cx gate on each pair of consecutive qubits

dcmckayibm commented on qiskit/ignis/verification/randomized_benchmarking/circuits.py:
```diff
@@ -143,7 +143,8 @@ def load_tables(max_nrb=2):
                                 length_multiplier=1, seed_offset=0,
-                                align_cliffs=False):
+                                align_cliffs=False,
+                                interleaved_gates=None):
```
These are going to get passed in as Clifford objects

ShellyGarion replied:
I think that it may be more difficult, since the function `compose_gates` in `clifford_utils.py` can only add gates to a Clifford object from a list of gates. In the current basic `Clifford.py` there is no function to compose two Clifford objects. 
I also think that for the users it may be easier to produce a list of gates than a Clifford object.

dcmckayibm replied:
Looking at the code 
1. `compose_gates` needs to be part of the clifford class
2. the clifford objects should be composable (this should be straightforward, let's ask @chriseclectic who understands the representation)

Then users input based on a clifford object (this way we know it's a clifford). The user can build that Clifford with the same list of gates so it's not really more work. The actual set of gates will get selected using the lookup table; that way it's in the same basis.

ShellyGarion replied:
The `Clifford.py` class should be a temporary class until there will be a Clifford python class in Terra, that will call the Aer Clifford simulator, so it includes only basic functionality.

As I remember, it was @chriseclectic who recommended which functions should be in the Clifford class, and which functions should be in `clifford_utils.py`  (based on the planned Clifford class).

If the `interleaved_gates` should be passed in as Clifford objects there are two possibilities: 

1. Add a function to the Clifford class that compose Clifford objects (is it basically just a multiplication of the two Symplectic matrices?)
**or:** 
2. Find the `index` of the Clifford object, translate it to canonical gates, and use the function `compose_gates`.

BTW, also when a Clifford is given as a list of gates we know it's a Clifford since it must use only the given Clifford gates (x,y,z,h,s,sdg,cx). 
So, I don't understand why it should be given as a Clifford object and not as Clifford gates.

dcmckayibm replied:
I agree we don't want to put too much effort into the `Clifford` class since it will get replaced by one in Terra, but I think these are quick fixes. Giving Clifford objects is more object-oriented and then we can do all error checking in the class.

ShellyGarion replied:
I discussed this with @chriseclectic .  He thinks that appending a compose method to the Clifford class is not immediate (since one has to be careful with updating the phases), and anyway this will be done in the new Clifford class in Terra.

So, we can either wait for the Terra Clifford class, or handle this for 1-2 qubits using an indirect way, namely: find the index of the Clifford object, translate it to canonical gates, and use the function compose_gates (this shouldn't be difficult).

dcmckayibm replied:
Ok, just pass in the list of gates...we'll wait for the Clifford class in terra (there will be a lot of updates for that anyway)

dcmckayibm commented:
For the fitter, I would plot both RB curves on the same plot and put the alpha from each and then the error of the interleaved gate

ShellyGarion commented:
@dcmckayibm - could you please review `interleaved_fitters.py` and the Jupyter notebook? 
Then I would be able to complete the notebook and write a test for the interleaved fitters.

ShellyGarion commented:
I think that the code is ready for review. I still need to:

- [ ] Write tests for the interleaved fitters.
- [ ] Replace the input from a list of Cliffrod gates to a Clifford object (waiting for Chris's advice - see the discussion above).
- [ ] Add more explanation to the Jupyter tutorial notebook.
- [ ] Add interleaved RB to the community RB overview notebook and to the documentation.

dcmckayibm commented:
> @dcmckayibm - could you please review `interleaved_fitters.py` and the Jupyter notebook?
> Then I would be able to complete the notebook and write a test for the interleaved fitters.

The plots are a little busy. I would drop the line connecting the points and replace it with a marker for each mean value. So the only lines would be the fit curves. Because there's two sets of points they need more differentiation. I would make the points and fit line the same color. There also needs to be a small legend to show which points are standard and which are interleaved. 

Don't list the estimated range on the label.

ShellyGarion commented:
> 
> 
> > @dcmckayibm - could you please review `interleaved_fitters.py` and the Jupyter notebook?
> > Then I would be able to complete the notebook and write a test for the interleaved fitters.
> 
> The plots are a little busy. I would drop the line connecting the points and replace it with a marker for each mean value. So the only lines would be the fit curves. Because there's two sets of points they need more differentiation. I would make the points and fit line the same color. There also needs to be a small legend to show which points are standard and which are interleaved.
> 
> Don't list the estimated range on the label.

Done.

ShellyGarion commented:
It seems that the RB fitters broke on the interleaved circuits, since the `circ_name=` `'rb_length_%d_seed_%d'` is hard coded in the RBFitter, and I wanted to give a different name to the interleaved rb circuits:  `'rb_interleaved_length_%d_seed_%d'`. 
My current solution is to change the RBFitter class, so that the hard coded circ_name will depend on a flag indicating whether this is an interleaved circuit.
@yaelbh made some more general suggestions in ignis internal (https://github.ibm.com/IBMQuantum/qiskit-ignis-internal/issues/58)

dcmckayibm commented:
> It seems that the RB fitters broke on the interleaved circuits, since the `circ_name=` `'rb_length_%d_seed_%d'` is hard coded in the RBFitter, and I wanted to give a different name to the interleaved rb circuits: `'rb_interleaved_length_%d_seed_%d'`.
> My current solution is to change the RBFitter class, so that the hard coded circ_name will depend on a flag indicating whether this is an interleaved circuit.
> @yaelbh made some more general suggestions in ignis internal (https://github.ibm.com/IBMQuantum/qiskit-ignis-internal/issues/58)

Why not just check if it has either of these names?

dcmckayibm commented on qiskit/ignis/verification/randomized_benchmarking/fitters.py:
```diff
@@ -35,7 +35,7 @@ class RBFitter:
 
     def __init__(self, backend_result, cliff_lengths,
-                 rb_pattern=None):
+                 rb_pattern=None, is_interleaved=False):
```
I think this can just be inferred by looking at the name

ShellyGarion replied:
Done

dcmckayibm commented on qiskit/ignis/verification/randomized_benchmarking/interleaved_fitters.py:
```diff
@@ -0,0 +1,274 @@
+# -*- coding: utf-8 -*-
```
This shouldn't be in a separate file, should be in fitters.py

ShellyGarion replied:
Done.

ShellyGarion commented:
> 
> 
> > It seems that the RB fitters broke on the interleaved circuits, since the `circ_name=` `'rb_length_%d_seed_%d'` is hard coded in the RBFitter, and I wanted to give a different name to the interleaved rb circuits: `'rb_interleaved_length_%d_seed_%d'`.
> > My current solution is to change the RBFitter class, so that the hard coded circ_name will depend on a flag indicating whether this is an interleaved circuit.
> > @yaelbh made some more general suggestions in ignis internal (https://github.ibm.com/IBMQuantum/qiskit-ignis-internal/issues/58)
> 
> Why not just check if it has either of these names?

You mean that I should give the interleaved circuits the same name as the original rb circuits? (I thought that it may be confusing)

dcmckayibm commented:
> > > It seems that the RB fitters broke on the interleaved circuits, since the `circ_name=` `'rb_length_%d_seed_%d'` is hard coded in the RBFitter, and I wanted to give a different name to the interleaved rb circuits: `'rb_interleaved_length_%d_seed_%d'`.
> > > My current solution is to change the RBFitter class, so that the hard coded circ_name will depend on a flag indicating whether this is an interleaved circuit.
> > > @yaelbh made some more general suggestions in ignis internal (https://github.ibm.com/IBMQuantum/qiskit-ignis-internal/issues/58)
> > 
> > 
> > Why not just check if it has either of these names?
> 
> You mean that I should give the interleaved circuits the same name as the original rb circuits? (I thought that it may be confusing)

You don't need the user to set a flag. If the circuit name starts with `rb_length` it's standard rb, if it starts with `rb_interleaved_length` it's an interleave circuit

dcmckayibm commented:
Can you also run a sanity check on the interleaved and make sure the values you get out correspond to the noise setting? So
1. Run interleaved
2. Run tomography on the interleaved gates

1 and 2 should agree (more or less)

ShellyGarion commented:
> 
> 
> Can you also run a sanity check on the interleaved and make sure the values you get out correspond to the noise setting? So
> 
>     1. Run interleaved
> 
>     2. Run tomography on the interleaved gates
> 
> 
> 1 and 2 should agree (more or less)

I added process tomography on the interleaved gates - see the Jupyter notebook "Interleaved_RB" in this PR - was that what you meant?

ShellyGarion commented:
@dcmckayibm - I think that the code is now complete for review. 
I addressed your previous comments and added:
- test_fitters.py now contains a test of InterleavedRBFitter
- the Jupyter notebook also contains process tomography on the interleaved gates
